code/controllers/subsystem/SStgui.dm

```diff
@@ -1,311 +1,361 @@
+		mob/user,
+		datum/src_object,
+		datum/tgui/ui,
+		force_open = FALSE)
+	// Look up a UI if it wasn't passed
+	if(isnull(ui))
+		ui = get_open_ui(user, src_object)
+	// Couldn't find a UI.
+	if(isnull(ui))
+		return null
+	var/data = src_object.ui_data(user) // Get data from the src_object.
+	if(force_open) // UI is already open; update it.
+		ui.send_full_update(data, TRUE)
+		return ui // We found the UI, return it
+	ui.process_status()
+	// UI ended up with the closed status
+	// or is actively trying to close itself.
+	// FIXME: Doesn't actually fix the paper bug.
+	if(ui.status <= UI_CLOSE)
+		ui.close()
+		return null
+	ui.send_update()
+	return ui
 
 /**
+ * public
+ *
+ * Get a open UI given a user and src_object.
  *
- * Gets the amount of open UIs on an object
- * Returns the number of UIs open.
+ * required user mob The mob who opened/is using the UI.
+ * required src_object datum The object/datum which owns the UI.
  *
- * * datum/src_object - The object/datum which owns the UIs.
+ * return datum/tgui The found UI.
  */
-/datum/controller/subsystem/tgui/proc/get_open_ui_count(datum/src_object)
-	if(!src_object.unique_datum_id) // First check if the datum has an UID set
-		return 0
-	var/src_object_key = "[src_object.UID()]"
-	if(isnull(open_uis[src_object_key]) || !islist(open_uis[src_object_key]))
-		return 0 // Couldn't find any UIs for this object.
-
-	var/open_count = 0
-	for(var/ui_key in open_uis[src_object_key])
-		for(var/datum/tgui/ui in open_uis[src_object_key][ui_key])
-			if(ui && ui.src_object && ui.user && ui.src_object.ui_host(ui.user)) // Check the UI is valid.
-				open_count++ // Count each UI thats open
+/datum/controller/subsystem/tgui/proc/get_open_ui(mob/user, datum/src_object)
+	var/key = "[src_object.UID()]"
+	// No UIs opened for this src_object
+	if(isnull(open_uis_by_src[key]) || !istype(open_uis_by_src[key], /list))
+		return null
+	for(var/datum/tgui/ui in open_uis_by_src[key])
+		// Make sure we have the right user
+		if(ui.user == user)
+			return ui
+	return null
 
-	return open_count
+/**
+ * public
+ *
+ * Update all UIs attached to src_object.
+ *
+ * required src_object datum The object/datum which owns the UIs.
+ *
+ * return int The number of UIs updated.
+ */
+/datum/controller/subsystem/tgui/proc/update_uis(datum/src_object)
+	var/count = 0
+	var/key = "[src_object.UID()]"
+	// No UIs opened for this src_object
+	if(isnull(open_uis_by_src[key]) || !istype(open_uis_by_src[key], /list))
+		return count
+	for(var/datum/tgui/ui in open_uis_by_src[key])
+		// Check if UI is valid.
+		if(ui && ui.src_object && ui.user && ui.src_object.ui_host(ui.user))
+			ui.process(force = TRUE)
+			count++
+	return count
 
+/**
+ * public
+ *
+ * Close all UIs attached to src_object.
+ *
+ * required src_object datum The object/datum which owns the UIs.
+ *
+ * return int The number of UIs closed.
+ */
+/datum/controller/subsystem/tgui/proc/close_uis(datum/src_object)
+	var/count = 0
```

Comment:
could instead store this value in `.` since `count` functionally is doing the same thing